lib/redis-pool.js

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const genericPool = require('generic-pool')
+const P = require('./promise')
+const redis = require('redis')
+const redisConnection = require('./redis-connection')
+
+P.promisifyAll(redis.RedisClient.prototype)
+P.promisifyAll(redis.Multi.prototype)
+
+module.exports = (config, log) => {
+  if (! config.redis.enabled) {
+    log.info({ op: 'redis.disabled' })
+    return
+  }
+
+  const redisConfig = {
+    host: config.redis.host,
+    port: config.redis.port,
+    prefix: config.redis.sessionsKeyPrefix,
+    // Prefer redis to fail fast than wait indefinitely for reconnection
+    enable_offline_queue: false
+  }
+
+  log.info({ op: 'redis.enabled', config: redisConfig })
+
+  const redisFactory = {
+    create () {
+      return new P(resolve => {
+        const client = redis.createClient(redisConfig)
+        client.on('ready', () => resolve(redisConnection(log, client)))
+        client.on('error', err => log.error({
+          op: 'redis.error',
+          err: err.message,
+          stack: err.stack
+        }))
+      })
+    },
+
+    destroy (client) {
+      return new P(resolve => {
+        client.quit()
+        client.on('end', resolve)
+      })
+    }
+  }
+
+  return genericPool.createPool(redisFactory, {
```

Comment:
May be worth adding a listener for `factoryCreateError` [1] here, if only to ensure any errors get logged.

[1] https://www.npmjs.com/package/generic-pool#poolon`

Reply:
Actually, reading those docs, it sounds like that error is only emitted if our own `factory.create` method rejects. But as it stands, it only resolves.

If we wanted to change that, we need to implement a retry strategy for connecting to Redis and fail after *n* attempts or seconds. This would mostly affect the behaviour of updates since our gets already just check whether `this.redis` is set before using it. But for updates, requests to the auth server will just sit there waiting to timeout if there's no connection to Redis. Happy to change this, what do we think would be a good value for *n*?

Reply:
Sorry, just realised I already answered my own question by including `acquireTimeoutMillis` in config. Presumably that already does what we want then. I'll have a play to make sure.

Reply:
(and implement a corresponding retry strategy so that we don't create zombie connections.

Reply:
Moved this back to the `active` column because this timeout stuff has made me realise there's a conflict between the way I'm treating reads and writes at the moment.

For reads, the rest of the code assumes it's fine for connecting to Redis to take however long it takes. Until that promise resolves we just don't bother getting data from Redis.

Obviously that doesn't work for writes, which is the entire rationale behind this PR. But I'm still using those semantics for `del`, which is wrong. We absolutely need `del` to fail when Redis is down or hasn't connected yet.

And because `generic-pool` doesn't provide a way to over-ride config options for individual calls to acquire, pursuing both approaches would actually require us to have to different connection pools, one set to timeout and the other not.

That seems pretty lame, it's much simpler just to have the same timeout semantics everywhere, so I'm switching to that.